### DISCUSSION TITLE
401 on Playbook play creation

### DISCUSSION THREAD

**Discussion by stijnpieters** [Security-Onion-Solutions/securityonion#1896] [Unsupported Versions] [answered]
Upon install of Security Onion 2.3.2, my Playbook was empty. Running so-playbook-ruleupdate only returned 'Play Creation Error -'.

Logs show 401 errors. Soctopus tries creating rules using the "anonymous" user, I'm assuming this user does not have rights to create new plays.

Upon further inspection I found that playbook.api_key was not set in /opt/so/saltstack/local/pillar/global.sls
When filling in that value with the api key from the admin user, and restarting soctopus, nothing changed. Still have 401 errors.

Any ideas what's going wrong?

Details of my install:
- Managersearch node
   - Listen on FQDN, not IP or hostname
   - all components installed
- multiple sensors 

EDIT:
Just restarting so-soctopus did not update SOCtopus.conf, the api_key in global.sls did not get parsed into the file using so-soctopus-restart. By stopping and starting using so-soctopus-start and -stop, SOCtopus.conf did update, and now plays can be created. 

New question: I'm assuming the api_key not being filled in is not intended behaviour. Any idea what caused it?
I did get errors at the end of my install (MySQL not starting), but these were fixed by running so-docker-refresh.

**Comment by kiaraho** [comment]
Same issue here on SOC 2.3.2. Ran `salt-call state.apply playbook.db_init`, followed by `salt-call state.apply playbook.automation_user_create` as suggested in #1592. I then did a `so-playbook-restart`, followed by a `so-playbook-ruleupdate`. Here is some of the output (truncated) I got on that last command:
```
Play Creation Error -
Play Creation Error -
Play Creation Error -

...

sigma/rules/windows/file_event/sysmon_powershell_exploit_scripts.yml
No Current Play - Create New Play in Playbook
{'play_creation': 401, 'play_url': 'failed'}
new

...

sigma/rules/windows/process_creation/win_apt_wocao.yml
Error - Sigma rule skipped 
expected a single document in the stream
  in "<unicode string>", line 1, column 1:
    action: global
    ^ (line: 1)
but found another document
  in "<unicode string>", line 14, column 1:
    ---
    ^ (line: 14)Play Creation Error - 

...

-= Update Summary =-

Sigma Community Repo:
 Already up to date.

Updated Plays: 0
New Plays: 495
No Updates Needed: 0

Enabled Rulesets:
['windows']



-= Completed: 2020-11-15 04:02:22.919081-=
```
I do not remember having Playbook issues on SOC 2.3.1.

**Reply by thejavagirl** [reply]
I'm having all the same issues with Playbook importing rules from sigma and adding new rules. I've reinstalled SO several times including the latest v2.3.10 without success. Getting nothing but 'Play Creation Error' when running so-playbook-ruleupdate and Error 401 when trying to add plays manually. My install is SO v2.3.10, Ubuntu 18.04LTS, Standalone on Quad-Core, 16GB Ram, 2TB SSD, Dedicated machine. I was dual booting Windows and Ubuntu but now I'm going to try a fresh install of SO and dedicate the entire machine to see if it makes any difference.

**Reply by defensivedepth** [reply]
@thejavagirl If you see this again, please open up a new Discussion about it - that would make it easier for us to track. Thanks!

**Reply by thejavagirl** [reply]
@defensivedepth - Will do. New install is running... Thanks for the fast reply!

**Reply by thejavagirl** [reply]
Dedicated install seems to have solved the issue for me after nothing else did. I'm guessing there's some conflict with Playbook working on Ubuntu 18.04? Not too sure but my issue is solved so I'm happy :) Thanks again for your help!

**Comment by weslambert** [comment, marked as answer]
This should be resolved in 2.3.10(available now).  Please try a fresh install and let me know if you get the same results.

**Reply by kiaraho** [reply]
Before I do a fresh install, is it possible to reinstall individual components such as `so-mysql` and `so-playbook`? I forgot to mention that the `salt-call state.apply playbook.automation_user_create` command above resulted in a `MySQL Error 1130: Host '10.48.0.10' is not allowed to connect to this MySQL server` error.

It would be helpful to know if individual components can be fixed without wiping everything with a fresh install. Right now it's not a big deal since this a test environment, but I can see how this would be an issue in a production environment.

**Reply by defensivedepth** [reply]
It depends on the component, but in general, yes, you should be able to fix / reset components without having to do a fresh install. We are still working on the tooling for some of this - for example, in this latest release (2.3.10), there is better setup stability for mysql (which is what Playbook depends on), and there is a new script called `so-playbook-reset` which will reset the entire component (database included) if needed.

**Reply by kiaraho** [reply]
@weslambert, I did a fresh Ubuntu 18 install and ran `so-setup-network`. The [sosetup.log](https://github.com/Security-Onion-Solutions/securityonion/files/5576898/sosetup.log) shows the same MySQL errors. `so-playbook-reset`generated the related errors below. Is is possible access MySQL directly and allow specified IP? It seems there are post-install scripts that should have done this, but failed to run.

```
Importing Plays - this will take some time....
/usr/sbin/so-playbook-reset: line 25: wait: pid 5 is not a child of this shell

-= Started: 2020-11-21 02:31:18.780817-=



-= Creating/Updating Plays based on the following categories: ['windows'] -=




-= Parsing current Plays in Playbook -=


Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/urllib3/connection.py", line 159, in _new_conn
    conn = connection.create_connection(
  File "/usr/local/lib/python3.9/site-packages/urllib3/util/connection.py", line 84, in create_connection
    raise err
  File "/usr/local/lib/python3.9/site-packages/urllib3/util/connection.py", line 74, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 670, in urlopen
    httplib_response = self._make_request(
  File "/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 392, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/local/lib/python3.9/http/client.py", line 1255, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/local/lib/python3.9/http/client.py", line 1301, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.9/http/client.py", line 1250, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.9/http/client.py", line 1010, in _send_output
    self.send(msg)
  File "/usr/local/lib/python3.9/http/client.py", line 950, in send
    self.connect()
  File "/usr/local/lib/python3.9/site-packages/urllib3/connection.py", line 187, in connect
    conn = self._new_conn()
  File "/usr/local/lib/python3.9/site-packages/urllib3/connection.py", line 171, in _new_conn
    raise NewConnectionError(
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f7796dd4af0>: Failed to establish a new connection: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 726, in urlopen
    retries = retries.increment(
  File "/usr/local/lib/python3.9/site-packages/urllib3/util/retry.py", line 446, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='10.48.0.10', port=3200): Max retries exceeded with url: /playbook/issues.json?offset=0&tracker_id=1&limit=100 (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f7796dd4af0>: Failed to establish a new connection: [Errno 111] Connection refused'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/SOCtopus/playbook_bulk-update.py", line 117, in <module>
    response = requests.get(url, headers=playbook_headers, verify=False).json()
  File "/usr/local/lib/python3.9/site-packages/requests/api.py", line 76, in get
    return request('get', url, params=params, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 530, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 643, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='10.48.0.10', port=3200): Max retries exceeded with url: /playbook/issues.json?offset=0&tracker_id=1&limit=100 (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f7796dd4af0>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

**Reply by defensivedepth** [reply]
@kiaraho What are the system specs for the Manager? CPU/Memory/ what kind of HDD?

**Reply by kiaraho** [reply]
Standalone Manager is a VirtualBox VM with 4 CPU cores, 16GB of memory and 200GB of storage. The host system has a 2TB SSD.

**Reply by defensivedepth** [reply]
@kiaraho What kind of HD? SSD? Mechanical, 5400/7200 rpm?

**Reply by kiaraho** [reply]
SAMSUNG 860 EVO Series 2.5" 2TB SATA III SSD

**Reply by defensivedepth** [reply]
@kiaraho Are there two IP's configured on any of the interfaces? Also, do any of the interfaces have DHCP instead of a static IP? 

Finally, can you please run `sudo salt-call grains.get ip_interfaces` 

Trying to run down a couple things.

**Reply by kiaraho** [reply]
@defensivedepth, the Standalone manager has three interfaces. A `enp0s3` interface on the 10.0.2.0/24 subnet to access the Internet, a `enp0s8` interface on the 10.48.0.0/24 subnet for Wazuh agents to connect to the manager, and a `enp0s9` monitoring interface on the 10.48.0.0/24 subnet. The first two interfaces have static IPs configured. The third interface is in promiscuous mode with no IP.

Here is the output of the `sudo salt-call grains.get ip_interfaces` command:

```
local:
    ----------
    bond0:
    docker0:
        - 172.17.0.1
    enp0s3:
        - 10.0.2.16
        - fe80::a00:27ff:fe19:e3b1
    enp0s8:
        - 10.48.0.10
        - fe80::a00:27ff:fe88:27d7
    enp0s9:
    lo:
        - 127.0.0.1
    veth349c2fa:
    veth6758818:
    veth90c30a4:
    vethaa849cb:
    vethbf9de9a:
    vethc85b5eb:
    vethd61e5ad:
    vethddf97f7:
```

**Reply by defensivedepth** [reply]
@kiaraho This is an unsupported configuration, which is why the install failed - The interface chosen as the Manager interface needs to be the default route for the system.

We are adding a pre-flight check for this.

**Comment by stijnpieters** [comment]
For me, the issue was basically resolved by updating to SO2.3.10.
My so-nginx was not restarted by 'soup', but running so-nginx-restart then fixed that and gave me the newer Navigator version from the 2.3.10 nginx image.
